apps/interpret_nn/util/error_util.h

```diff
@@ -40,43 +37,53 @@ inline std::ostream &operator<<(std::ostream &s, const std::vector<T> &v) {
     return s << "}";
 }
 
+// Note: all severity values output to stderr, not stdout.
+// Note: ERROR does *not* trigger an exit()/abort() call.
+enum LogSeverity {
+    INFO = 0,
+    WARNING = 1,
+    ERROR = 2,
+};
+
 namespace internal {
 
-struct FatalError final {
+struct Logger {
     std::ostringstream msg;
+    const LogSeverity severity;
 
-    FatalError(const char *file, int line, const char *condition_string) {
-        msg << "Error @ " << file << ":" << line << ".";
-        if (condition_string) {
-            msg << " Condition failed: " << condition_string;
-        }
-        msg << "\n";
-    }
+    Logger(LogSeverity severity);
+    Logger(LogSeverity severity, const char *file, int line);
 
-    [[noreturn]] ~FatalError() noexcept(false) {
-        if (!msg.str().empty() && msg.str().back() != '\n') {
-            msg << "\n";
-        }
-
-        std::cerr << msg.str();
-        abort();
-    }
+    ~Logger() noexcept(false);
 
     template<typename T>
-    FatalError &operator<<(const T &x) {
+    Logger &operator<<(const T &x) {
         msg << x;
         return *this;
     }
 
-    FatalError &ref() {
+    Logger &ref() {
         return *this;
     }
 
-    FatalError() = delete;
-    FatalError(const FatalError &) = delete;
-    FatalError &operator=(const FatalError &) = delete;
-    FatalError(FatalError &&) = delete;
-    FatalError &operator=(FatalError &&) = delete;
+    Logger() = delete;
+    Logger(const Logger &) = delete;
+    Logger &operator=(const Logger &) = delete;
+    Logger(Logger &&) = delete;
+    Logger &operator=(Logger &&) = delete;
+};
+
+struct CheckLogger : public Logger {
+    CheckLogger(LogSeverity severity, const char *condition_string);
+    CheckLogger(LogSeverity severity, const char *file, int line, const char *condition_string);
+
+    [[noreturn]] ~CheckLogger() noexcept(false);
+
+    CheckLogger() = delete;
```

Comment:
Don't think you need to do this if the base class version is deleted? Not sure though.